### PR TITLE
Feature ap 243/v3 client

### PR DESCRIPTION
**Pull request by ccleung**
Tracker Ticket:
* https://quandl.atlassian.net/browse/AP-352
* https://quandl.atlassian.net/browse/AP-1624 from: https://github.com/quandl/quandl-python/pull/42

## Description:  
Updated python package, uses v3 api

## Note to the reviewer:
for commit https://github.com/quandl/quandl-python/commit/d2034eb80a6cbb412e2047d7ec682adef4bfa2db:
* pass download_type as a query param to bulk download methods
* support both folder path and filename for bulk download
* updated tests and readme

## Note to QA:
* Ensure we test pandas/numpy data is expected
* Ensure data is returned in ascending order by default, and descending order if order type is specified
* Ensure README is clear and steps all work as expected
* Ensure api key and api version configurations work as expectged

### CODE REVIEW

File: quandl/model/database.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        url = Util.constructed_path(url, {'id': self.database_code})
+
+        params = {}
+        if 'download_type' in options:
+            params['download_type'] = options['download_type']
+        if ('include_key' not in options or options['include_key']) \
+                and ApiConfig.api_key:
+            params['auth_token'] = ApiConfig.api_key
+
+        if list(params.keys()).__len__() > 0:
```
Comment (ccleung) [inline quandl/model/database.py]: maybe change to len()

File: quandl/model/database.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            url += '?' + urlencode(params)
+
+        return url
+
+    def bulk_download_to_file(self, folder_path, **options):
```
Comment (ccleung) [inline quandl/model/database.py]: make sure to document these methods

File: quandl/model/dataset.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+class Dataset(GetOperation, ListOperation, ModelBase):
+
+    @classmethod
+    def get_path(cls):
+        return "%s/metadata" % cls.default_path()
+
+    def data(self, **options):
+        # default order to ascending, and respect whatever user passes in
+        params = {'database_code': self.database_code,
```
Comment (ccleung) [inline quandl/model/dataset.py]: consider not allowing user to override these attributes

File: quandl/model/merged_dataset.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+    def _rename_columns(self, dataset_metadata, original_column_name):
+        return "%s/%s" % (dataset_metadata.database_code,
+                          dataset_metadata.dataset_code) + ' - ' + original_column_name
+
+    def _get_dataset_attribute(self, k):
+        return list(unique_everseen([dataset.raw_data[k] for dataset in self._datasets]))
+
+    def _initialize_raw_data(self):
+        raw_data = self._datasets[0].raw_data.copy()
```
Comment (ccleung) [inline quandl/model/merged_dataset.py]: check for size

File: quandl/util.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    @staticmethod
+    def methodize(string):
+        return parameterize(string.replace(r'.', '')).replace(r'-', '_')
+
+    @staticmethod
+    def merge_options(key, dic, **options):
+        if key in options:
+            # respect values in options over hash
+            dic.update(options[key])
```
Comment (ccleung) [inline quandl/util.py]: maybe call merge_to_dicts

File: quandl/model/merged_dataset.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    @property
+    def oldest_available_date(self):
+        return min(self._get_dataset_attribute('oldest_available_date'))
+
+    @property
+    def newest_available_date(self):
+        return max(self._get_dataset_attribute('newest_available_date'))
+
+    def data(self, **options):
+        dataset_data_list = [dataset.data(**options)
```
Comment (ccleung) [inline quandl/model/merged_dataset.py]: make a copy for each call

File: quandl/model/merged_dataset.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+        for index, data_frame in enumerate(data_frames):
+            metadata = self._datasets[index]
+            data_frame.rename(
+                columns=lambda x: self._rename_columns(metadata, x), inplace=True)
+            merged_data_frames = pd.merge(
+                merged_data_frames, data_frame, right_index=True, left_index=True, how='outer')
+        # meta should be the same for every individual Dataset
+        # request, just take the first one
+        data_metadata = dataset_data_list[0].meta.copy()
```
Comment (ccleung) [inline quandl/model/merged_dataset.py]: rename top merged_data_metadata

File: quandl/operations/get.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+class GetOperation(Operation):
+    @classmethod
+    def get(cls, id, **options):
+        params = {'id': str(id)}
+        options = Util.merge_options('params', params, **options)
+        path = Util.constructed_path(cls.get_path(), options['params'])
+        r = Connection.request('get', path, **options)
+        response_data = r.json()
+        Util.convert_to_dates(response_data)
+        data = response_data[singularize(cls.lookup_key())]
```
Comment (ccleung) [inline quandl/operations/get.py]: consider in the future how to refactor these for Operations

File: quandl/model/merged_dataset.py (hunk 4 above)
Comment (ccleung) [inline quandl/model/merged_dataset.py]: done

File: quandl/model/dataset.py (hunk 3 above)
Comment (ccleung) [inline quandl/model/dataset.py]: needs more consideration, there is arguments for either case, and ruby client already allows user to override

File: quandl/model/database.py (hunk 1 above)
Comment (ccleung) [inline quandl/model/database.py]: updated

File: quandl/util.py (hunk 5 above)
Comment (ccleung) [inline quandl/util.py]: updated

File: quandl/model/merged_dataset.py (hunk 6 above)
Comment (ccleung) [inline quandl/model/merged_dataset.py]: actually it looks like options is always copied cause it is passed in as **options, no need to make a copy of each options, it is already done internally

File: quandl/model/merged_dataset.py (hunk 7 above)
Comment (ccleung) [inline quandl/model/merged_dataset.py]: updated

File: README.md [hunk 9]
```diff
@@ -1,189 +1,325 @@
...
 
 ```python
-Quandl.push(data, code='TEST', name='Test', desc='test')
+import quandl
+merged_dataset = quandl.MergedDataset.get([('GOOG/NASDAQ_AAPL', {'column_index': [4]}),
+                                           ('GOOG/NASDAQ_MSFT', {'column_index': [1,4]}), 'GOOG/NYSE_TWTR'])
 ```
 
-All parameters but desc are necessary
+In the above example, column 4 of dataset 'GOOG/NASDAQ_AAPL', columns 1 and 4 of dataset 'GOOG/NASDAQ_MSFT', and all columns of 'GOOG/NYSE_TWTR' is requested.
```
Comment (mbasset) [inline README.md]: This sentence reads a bit strangely. `In the above example, column 4 of dataset 'GOOG/NASDAQ_AAPL', columns 1 and 4 of dataset 'GOOG/NASDAQ_MSFT', and all columns of 'GOOG/NYSE_TWTR' is requested.`

Perhaps try:

In the above example the following data will be merged together:

* column 4 of dataset 'GOOG/NASDAQ_AAPL'
* columns 1 and 4 of dataset 'GOOG/NASDAQ_MSFT'
* and all columns of 'GOOG/NYSE_TWTR'

**Comment by ccleung** [comment]
Everything is good to go, waiting for email to be sent out to python users, as well as final round of testing for this branch.

For reference, to install:
`pip install https://github.com/quandl/quandl-python/archive/feature-AP-243/v3-client.zip`

File: quandl/model/database.py (hunk 2 above)
Comment (ccleung) [inline quandl/model/database.py]: updated README

**Comment by ccleung** [comment]
We need to add request_source: 
https://github.com/quandl/quandl-python/blob/master/Quandl/Quandl.py#L232

**Pull request merged**